Source/Core/VideoCommon/TextureCacheBase.cpp

```diff
@@ -237,22 +237,43 @@ TextureCacheBase::TCacheEntryBase* TextureCacheBase::DoPartialTextureUpdates(Tex
 		{
 			if (entry->hash == entry->CalculateHash())
 			{
+				MathUtil::Rectangle<int> srcrect, dstrect;
+
 				u32 block_offset = (entry->addr - entry_to_update->addr) / block_size;
 				u32 block_x = block_offset % numBlocksX;
 				u32 block_y = block_offset / numBlocksX;
 
-				u32 x = block_x * block_width;
-				u32 y = block_y * block_height;
-				MathUtil::Rectangle<int> srcrect, dstrect;
-				srcrect.left = 0;
-				srcrect.top = 0;
-				dstrect.left = 0;
-				dstrect.top = 0;
+				u32 dst_x = block_x * block_width;
+				u32 dst_y = block_y * block_height;
+				u32 src_x = 0;
+				u32 src_y = 0;
+
+				// If the EFB copy doesn't fully fit, cancel the copy process
+				if (entry->native_width - src_x > entry_to_update->native_width - dst_x
+					|| entry->native_height - src_y > entry_to_update->native_height - dst_y)
+				{
+					iter++;
+					continue;
+				}
+
+				u32 copy_width = std::min(entry->native_width - src_x, entry_to_update->native_width - dst_x);
+				u32 copy_height = std::min(entry->native_height - src_y, entry_to_update->native_height - dst_y);
+
+				if (g_ActiveConfig.bCopyEFBScaled)
+				{
+					src_x = Renderer::EFBToScaledX(src_x);
+					src_y = Renderer::EFBToScaledY(src_y);
+					dst_x = Renderer::EFBToScaledX(dst_x);
+					dst_y = Renderer::EFBToScaledY(dst_y);
+					copy_width = Renderer::EFBToScaledX(copy_width);
+					copy_height = Renderer::EFBToScaledY(copy_height);
+				}
+
 				if (entry_need_scaling)
 				{
 					entry_need_scaling = false;
-					u32 w = entry_to_update->native_width * entry->config.width / entry->native_width;
-					u32 h = entry_to_update->native_height * entry->config.height / entry->native_height;
+					u32 w = Renderer::EFBToScaledX(entry_to_update->native_width);
```

Comment:
I'm not entirely sure if it can recreate the same fractonial scaling the efb copies get.